MetricRegistries::registerOrReplace logs the metric, even if the metric itself isn't loggable

I'm seeing some pretty odd errors come from Metrics which seem related to the fact that the metric register logs the metric which it failed to register, even if the metric that failed to register is a lambda.

```
2019-05-08 11:20:19,441 Log4j2-TF-1-AsyncLogger[witchcraft]-1 ERROR SLS-Logging: Failed to json encode an object of type [com.palantir.tritium.metrics.MetricRegistries$$Lambda$113/1911138454] com.fasterxml.jackson.databind.JsonMappingException: Failed to load class 'com.palantir.tritium.metrics.MetricRegistries$$Lambda$113/1911138454$Access4JacksonSerializer9adef09f': IllegalName: com.palantir.tritium.metrics.MetricRegistries$$Lambda$113/1911138454$Access4JacksonSerializer9adef09f
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:292)
	at com.fasterxml.jackson.databind.SerializerProvider.reportMappingProblem(SerializerProvider.java:1223)
	at com.fasterxml.jackson.databind.SerializerProvider._createAndCacheUntypedSerializer(SerializerProvider.java:1341)
	at com.fasterxml.jackson.databind.SerializerProvider.findValueSerializer(SerializerProvider.java:510)
	at com.fasterxml.jackson.databind.SerializerProvider.findTypedValueSerializer(SerializerProvider.java:713)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:308)
	at com.fasterxml.jackson.databind.ObjectWriter$Prefetch.serialize(ObjectWriter.java:1396)
	at com.fasterxml.jackson.databind.ObjectWriter._configAndWriteValue(ObjectWriter.java:1120)
	at com.fasterxml.jackson.databind.ObjectWriter.writeValue(ObjectWriter.java:966)
	at com.palantir.sls.logging.log4j.ServiceLogEncoder.lambda$static$4(ServiceLogEncoder.java:71)
	at com.palantir.sls.logging.log4j.Log4jEventAdapter.lambda$static$2(Log4jEventAdapter.java:45)
	at com.palantir.sls.logging.log4j.Log4jEventAdapter.visitParameters(Log4jEventAdapter.java:368)
	at com.palantir.sls.logging.log4j.Log4jEventAdapter.appendParams(Log4jEventAdapter.java:159)
	at com.palantir.sls.logging.log4j.Log4jEventAdapter.appendParams(Log4jEventAdapter.java:27)
	at com.palantir.sls.logging.log4j.ServiceLogEncoder.appendEvent(ServiceLogEncoder.java:131)
	at com.palantir.sls.logging.log4j.ServiceLogLayout.encode(ServiceLogLayout.java:66)
	at com.palantir.sls.logging.log4j.ServiceLogLayout.encode(ServiceLogLayout.java:21)
	at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.directEncodeEvent(AbstractOutputStreamAppender.java:197)
	at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.tryAppend(AbstractOutputStreamAppender.java:190)
	at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.append(AbstractOutputStreamAppender.java:181)
	at org.apache.logging.log4j.core.appender.RollingRandomAccessFileAppender.append(RollingRandomAccessFileAppender.java:252)
	at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(AppenderControl.java:156)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppender0(AppenderControl.java:129)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(AppenderControl.java:120)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppender(AppenderControl.java:84)
	at org.apache.logging.log4j.core.config.LoggerConfig.callAppenders(LoggerConfig.java:464)
	at org.apache.logging.log4j.core.config.LoggerConfig.processLogEvent(LoggerConfig.java:448)
	at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:431)
	at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:419)
	at org.apache.logging.log4j.core.config.DefaultReliabilityStrategy.log(DefaultReliabilityStrategy.java:60)
	at org.apache.logging.log4j.core.async.AsyncLogger.actualAsyncLog(AsyncLogger.java:381)
	at org.apache.logging.log4j.core.async.RingBufferLogEvent.execute(RingBufferLogEvent.java:161)
	at org.apache.logging.log4j.core.async.RingBufferLogEventHandler.onEvent(RingBufferLogEventHandler.java:45)
	at org.apache.logging.log4j.core.async.RingBufferLogEventHandler.onEvent(RingBufferLogEventHandler.java:29)
	at com.lmax.disruptor.BatchEventProcessor.processEvents(BatchEventProcessor.java:168)
	at com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:125)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: Failed to load class 'com.palantir.tritium.metrics.MetricRegistries$$Lambda$113/1911138454$Access4JacksonSerializer9adef09f': IllegalName: com.palantir.tritium.metrics.MetricRegistries$$Lambda$113/1911138454$Access4JacksonSerializer9adef09f
	at com.fasterxml.jackson.module.afterburner.util.MyClassLoader.loadAndResolve(MyClassLoader.java:100)
	at com.fasterxml.jackson.module.afterburner.ser.PropertyAccessorCollector.generateAccessorClass(PropertyAccessorCollector.java:170)
	at com.fasterxml.jackson.module.afterburner.ser.PropertyAccessorCollector.findAccessor(PropertyAccessorCollector.java:97)
	at com.fasterxml.jackson.module.afterburner.ser.SerializerModifier.changeProperties(SerializerModifier.java:63)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.constructBeanSerializer(BeanSerializerFactory.java:381)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.findBeanSerializer(BeanSerializerFactory.java:279)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory._createSerializer2(BeanSerializerFactory.java:231)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.createSerializer(BeanSerializerFactory.java:165)
	at com.fasterxml.jackson.databind.SerializerProvider._createUntypedSerializer(SerializerProvider.java:1388)
	at com.fasterxml.jackson.databind.SerializerProvider._createAndCacheUntypedSerializer(SerializerProvider.java:1336)
	... 34 more
Caused by: java.lang.NoClassDefFoundError: IllegalName: com.palantir.tritium.metrics.MetricRegistries$$Lambda$113/1911138454$Access4JacksonSerializer9adef09f
	at java.lang.ClassLoader.preDefineClass(ClassLoader.java:654)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:761)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:642)
	at com.fasterxml.jackson.module.afterburner.util.MyClassLoader.loadAndResolve(MyClassLoader.java:94)
```

If I dive in, this error comes from:

```
        registerSafe(metrics, MetricRegistry.name(MetricRegistries.class.getPackage().getName(), "snapshot", "now"),
                (Gauge<String>) MetricRegistries::nowIsoTimestamp);
```

Where `MetricRegistries::nowIsoTimestamp` is just a lambda, so the Jackson afterburner serialization fails

--- a/tritium-metrics/src/main/java/com/palantir/tritium/metrics/MetricRegistries.java
+++ b/tritium-metrics/src/main/java/com/palantir/tritium/metrics/MetricRegistries.java
@@ -370,13 +370,15 @@ public final class MetricRegistries {
                 if (replace && registry.remove(name)) {
                     logger.info("Removed existing registered metric with name {}: {}",
                             SafeArg.of("name", name),
-                            SafeArg.of("existingMetric", existingMetric));
+                            // #256: Metric implementations are necessarily json serializable
+                            SafeArg.of("existingMetric", String.valueOf(existingMetric)));
                     registry.register(name, metric);
                     return metric;
                 } else {
                     logger.warn("Metric already registered at this name. Name: {}, existing metric: {}",
                             SafeArg.of("name", name),
-                            SafeArg.of("existingMetric", existingMetric));
+                            // #256: Metric implementations are necessarily json serializable
+                            SafeArg.of("existingMetric", String.valueOf(existingMetric)));
                     @SuppressWarnings("unchecked")
                     T registeredMetric = (T) existingMetric;
                     return registeredMetric;

--- a/tritium-metrics/src/main/java/com/palantir/tritium/metrics/MetricRegistries.java
+++ b/tritium-metrics/src/main/java/com/palantir/tritium/metrics/MetricRegistries.java
@@ -362,9 +362,10 @@ public final class MetricRegistries {
                 Set<Class<?>> existingMetricInterfaces = ImmutableSet.copyOf(existingMetric.getClass().getInterfaces());
                 Set<Class<?>> newMetricInterfaces = ImmutableSet.copyOf(metric.getClass().getInterfaces());
                 if (!existingMetricInterfaces.equals(newMetricInterfaces)) {
-                    throw new IllegalArgumentException(
-                            "Metric already registered at this name that implements a different set of interfaces."
-                                    + " Name: " + name + ", existing metric: " + existingMetric);
+                    throw new SafeIllegalArgumentException(
+                            "Metric already registered at this name that implements a different set of interfaces",
+                            SafeArg.of("name", name),
+                            SafeArg.of("existingMetric", String.valueOf(existingMetric)));
                 }
 
                 if (replace && registry.remove(name)) {

--- a/tritium-metrics/src/test/java/com/palantir/tritium/metrics/MetricRegistriesTest.java
+++ b/tritium-metrics/src/test/java/com/palantir/tritium/metrics/MetricRegistriesTest.java
@@ -327,8 +327,8 @@ public class MetricRegistriesTest {
                 MetricRegistries.registerSafe(metrics, "test", metrics.histogram("histogram")))
                 .isInstanceOf(IllegalArgumentException.class)
                 .hasMessageStartingWith(
-                        "Metric already registered at this name that implements a different set of interfaces. "
-                                + "Name: test, existing metric: com.codahale.metrics.Counter");
+                        "Metric already registered at this name that implements a different set of interfaces: "
+                                + "{name=test, existingMetric=com.codahale.metrics.Counter");
     }
 
     @Test

